### PR TITLE
Add support for byref generic parameters in methods

### LINKED ISSUE #73: Method not found: 'Void System.__Canon..ctor(IntPtr)
MelonLoader v0.6.0 ALPHA Pre-Release
Game Name: TheLongDark
Unity Version: 2021.3.6f1
Game Version: 2.06
Using Il2CppInterop Version = 1.4.3-ci.244

The following exception occurs when trying to either
```
string text = "";
Il2CppTLD.Serialization.SerializationUtils.TryDeserializeObject<string>("some test text here", out text);
-----------------------------------------------
string text = "";
SaveGameSlots.TryLoadDataFromSlot<string>(SaveSlotName,"KeyToGet", out text);
```
```
System.MissingMethodException: Method not found: 'Void System.__Canon..ctor(IntPtr)'.
   at Il2CppTLD.Serialization.SerializationUtils.TryDeserializeObject[T](String text, T& result)
```

It was suggested to me that it's a potential problem with **new T(intPtr2)** in
```
[CallerCount(6)]
        [CachedScanResults(RefRangeStart = 1397578, RefRangeEnd = 1397584, XrefRangeStart = 1397567, XrefRangeEnd = 1397578, MetadataInitTokenRva = 0L, MetadataInitFlagRva = 0L)]
        public unsafe static bool TryDeserializeObject<T>(string text, out T result)
        {
            System.IntPtr* ptr = stackalloc System.IntPtr[2];
            *ptr = IL2CPP.ManagedStringToIl2Cpp(text);
            byte* num = (byte*)ptr + checked((nuint)1u * unchecked((nuint)sizeof(System.IntPtr)));
            System.IntPtr intPtr = IL2CPP.Il2CppObjectBaseToPtr(System.Runtime.CompilerServices.Unsafe.As<T, Il2CppObjectBase>(ref result));
            *(System.IntPtr**)num = &intPtr;
            System.IntPtr exc;
            System.IntPtr obj = IL2CPP.il2cpp_runtime_invoke(MethodInfoStoreGeneric_TryDeserializeObject_Public_Static_Boolean_String_byref_T_0<T>.Pointer, (System.IntPtr)0, (void**)ptr, ref exc);
            Il2CppException.RaiseExceptionIfNecessary(exc);
            System.IntPtr intPtr2 = intPtr;
            System.Runtime.CompilerServices.Unsafe.As<T, object>(ref result) = ((intPtr2 == (System.IntPtr)0) ? ((T)null) : new T(intPtr2));
            return *(bool*)IL2CPP.il2cpp_object_unbox(obj);
        }
```
"System.__Canon is an implementation detail related to the way generics are implemented in the CLR"
https://stackoverflow.com/questions/16854393/strange-system-canon-exception

Please let me know if i can provide any other information or test anything,
Dom (STBlade)

### PR DESCRIPTION
**Pull request by Kasuromi**
Adds support for byref generic parameters in methods, e.g. `Il2CppSystem.Collections.Dictionary.TryGetValue(TKey key, out TValue value)`

Closes #73

### CODE REVIEW

File: Il2CppInterop.Runtime/IL2CPP.cs [hunk 1]
```diff
@@ -287,9 +287,20 @@ public static void ThrowIfNull(object arg)
...
+            {
+                // NOTE(Kas): This assumes that objectPointer isn't pointing to the data, but is the actual data
+                // which has only started ocurring since adding support for generic byrefs in method proxygen
+                return Unsafe.As<IntPtr, T>(ref objectPointer);
+            }
+            else
+            {
+                objectPointer = il2cpp_value_box(Il2CppClassPointerStore<T>.NativeClassPtr, objectPointer);
+            }
+        }
```
Comment (js6pak) [inline Il2CppInterop.Runtime/IL2CPP.cs:303]: Is this change still necessary after 507a5ac30171b49df479e314c5ada772fee60f6b?

**Review by js6pak — CHANGES_REQUESTED** [review]

File: Il2CppInterop.Runtime/IL2CPP.cs (hunk 1 above)
Comment (Kasuromi) [reply Il2CppInterop.Runtime/IL2CPP.cs]: Good catch, it's not necessary - I just got rid of it.

**Review by js6pak — APPROVED** [review]
